[MOS-281] Base documentation update (#3645)
mudita/MuditaOS: doc/quickstart.md

            * [Running on linux](#running-on-linux)
            * [Running on rt1051 from emmc](#running-on-rt1051-from-emmc)
      * [checking commits](#checking-commits)
      * [CMake options](#cmake-options)
         * [Catching logs using UART](#catching-logs-using-uart)
   * [More details on preparing your local environment](#more-details-on-preparing-your-local-environment)
      * [Code style with git hooks](#code-style-with-git-hooks)
      * [Commit message template](#commit-message-template)
      * [Commit message hook](#commit-message-hook)
         * [Adding a hook:](#adding-a-hook)
         * [Using a hook:](#using-a-hook)
   * [Build using Docker](#build-using-docker)
      * [Building your own Docker image](#building-your-own-docker-image)
   * [Preparing packages](#preparing-packages)

## Introduction

The project can be build:
- on linux
- on windows with [WSL](https://docs.microsoft.com/en-us/windows/wsl/)
- with docker image
- Mac native compilation **is not supported**

### Download repository and submodules

MuditaOS uses submodules, therefore for clone use:
```
git clone --recurse-submodules git@github.com:mudita/MuditaOS.git
```

> if you didin't use `--recurse-submodules` or have to update submodules use:

Reviewer: ```suggestion
if you didn't use `--recurse-submodules` or have to update submodules use:
```

```
git submodule update --init --recursive
```

### Install dependencies

#### bootstrap

To install all the basic depenencies run: 
```
cd config && ./bootstrap.sh 0-
```

The script is written for Ubuntu and tested on 20.04.  

> **Note** that this:

Reviewer: ```suggestion
**Note** The script:
```

> - the script require sudo to:
>     - downloads a shellscript from `get.docker.com` and runs it with sudo for Docker installation.
>     - instals few required packages (list them with: `cat config/bootstrap_config`) which also require root privileges.
> - installs `GCCv10` and `CMake` to `$HOME` directory.
> - it's needed to be run only once

Reviewer: ```suggestion
-requires sudo to:
    - downloads a shell script from `get.docker.com` and runs it with sudo for Docker installation.
    - installs a few required packages (list them with: `cat config/bootstrap_config`) which also require root privileges.
- installs `GCCv10` and `CMake` to `$HOME` directory.
- it's needed to be run only once
```

After running provisioning you are ready to checkout and build project for both platforms supported.

You can see this `help` if you do not pass any param to `bootstrap.sh`:
```
[I] ⋊> ~/w/m/M/config on master ⨯ ./bootstrap.sh                                                                                                                                                                                                                                    09:00:01
test_if_run_as_root
Available build steps:
	0 install_hooks
	1 add_ignore_revs_for_blame
	2 install_ubuntu_packages
	3 setup_arm_toolchain
	4 setup_cmake
	5 setup_gcc_alternatives
	6 add_to_path ARM_GCC /home/pholat/gcc-arm-none-eabi-10-2020-q4-major/bin
	7 add_to_path CMAKE /home/pholat/cmake-3.21.3-linux-x86_64/bin
	8 install_docker
	9 add_to_docker_group
call:
./bootstrap.sh <build step no>[-]
ex.:
./bootstrap.sh 1       - build step 1 (add_ignore_revs_for_blame)
./bootstrap.sh 3       - build step 3 (setup_arm_toolchain)
./bootstrap.sh 3-      - build from step 3 to the end
./bootstrap.sh 0-      - build everything
```

#### JLink software

> We are using J-Link driver in version J-Link v634f ([Ubuntu download](https://www.segger.com/downloads/jlink/JLink_Linux_V634f_x86_64.deb))  

Reviewer: 698e is the last working version ;)

**NOTE:** This step is only required to load software and debug it via J-Link. It's not available for community - as it requires a programator board and soldered in programmator socket.

**WARNING:** newer J-Link software revisions seem to work fine, but loading via them do not result in usable software on the hardware tharget.

#### Add GitHub token

We are storing assets on github pages and these are downloaded automatically before software images are built.
Please follow github token configuration here: [download assets documentation](../doc/download_asset.py)

#### Project configuration

To configure project we have helper script: `./configure.sh` which essentially passes through basic cmake configuration for each product for each platform.

> We hihly advice to use `ninja` as it has proven better compilation times over make.

Reviewer: ```suggestion
We highly advise using `ninja` as it has proven better compilation times over make.
```

which can be run with following parameters:
```
# command      # product              # platform     # build type
./configure.sh [PurePhone|BellHybrid] [rt1051|linux] [release|debug|relwithdebinfo] [additional cmake parameters and options go here, i.e. -DENABLE_DEVELOPER_MODE_ENDPOINT=1 or "-G Ninja"]
```

__Examples:__

1. building pure phone for rt1051 platform with ninja

**NOTE** Due to software size we cant have full debug builds for rt1051 targets. Please use relwithdebinfo

```
./configure.sh pure rt1051 relWithDebInfo "-G Ninja"
```

> 2. building pure phone for linux platform with make

Reviewer: ```suggestion
2. building PurePhone for Linux platform with make
```

```
./configure.sh pure linux debug
```

3. building bell for linux platform with ninja

```
./configure.sh bell linux debug "-G Ninja"
```

#### Project build

> Each run of `configure.sh` created `build-{PRODUCT}-{PLATFORM}-{OPTMALIZAION}` folder, i.e.: `build-PurePhone-rt1051-RelWithDebInfo`

Reviewer: ```suggestion
Each run of `configure.sh` creates `build-{PRODUCT}-{PLATFORM}-{OPTMALIZAION}` folder, i.e.: `build-PurePhone-rt1051-RelWithDebInfo`
```

1. enter your selected build catalog
2. run `ninja` or `make` depending on your selection during configuration.
3. This will compile your selected target and create image of it

To know more about build targets please see: [build targets documentation](../doc/build_targets.md)

> **WARNING:** our source code is open source, but the project itself is in progress of fully embracing the community. Currently you can:
> 1. build binaries from the software
> 2. are not able to create images - due difficulties with separation of 3rd party propritetary assets

Reviewer: ```suggestion
**WARNING:** our source code is open source, but the project itself is in the progress of fully embracing the community. Currently you:
1. can build binaries from the software
2. are not able to create images - due to difficulties with separation of 3rd party proprietary assets
```

**Note** You can install and use ccache to speed up compilations

#### Project load and run

> Project load and run is platform dependent. Curently we suport only linux and rt1051 targets.

Reviewer: ```suggestion
Project load and run are platform-dependent. Currently, we support only Linux and rt1051 targets.
```

> **NOTE:** Addition of open source assets is in progress and should be done in following weeks. Till then Open source community cant build their own images - due to lack od fonts and luts. Support will be added step by step:

Reviewer: ```suggestion
**NOTE:** Addition of open source assets is in progress and should be done in the following weeks. Till then Open source community can't build their own images - due to a lack of fonts and LUTs. Support will be added step by step:
```

- addition of open source fonts
- addition of partial update packages (not signed) for Open Source contributors

##### Running on linux

Build default target and run `./PurePhone.elf` for phone target, or for BellHybrid: `./BellHybrid.elf`  

##### Running on rt1051 from emmc

Please follow: [running on rt1051 platform](./running_on_phone.md)

### checking commits

> All commits have to compy with checks with:

Reviewer: ```suggestion
All commits have to comply with checks with:
```

1. our pre commit hook
    1.1 required code have to be copyrighted with Mudita licensing
    1.2 required code have to comply to clang-format defined in repository
    1.3 code is checked against addition of binary blobs polluting software

> 2. pre merge CI job checking if commit is in acceptable format:

Reviewer: ```suggestion
2. pre-merge CI job checking if the commit is in acceptable format:
```
